gax/src/main/java/com/google/api/gax/tracing/OpencensusTracer.java

```diff
@@ -225,61 +225,68 @@ public void connectionSelected(int id) {
-    currentAttemptId = attemptNumber;
-    attemptSentMessages.set(0);
-    attemptReceivedMessages = 0;
-
-    HashMap<String, AttributeValue> attributes = new HashMap<>();
-    populateAttemptNumber(attributes);
-
-    span.addAnnotation("Attempt started", attributes);
-  }
-
   /** {@inheritDoc} */
   @Override
   public void attemptSucceeded() {
-    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    Map<String, AttributeValue> attributes = new HashMap<>();
 
-    span.addAnnotation("Attempt succeeded", attributes);
+    attemptCompleted("Attempt succeeded", attributes);
   }
 
   @Override
   public void attemptCancelled() {
-    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    Map<String, AttributeValue> attributes = new HashMap<>();
 
-    span.addAnnotation("Attempt cancelled", attributes);
+    attemptCompleted("Attempt cancelled", attributes);
   }
 
   /** {@inheritDoc} */
   @Override
   public void attemptFailed(Throwable error, Duration delay) {
-    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    Map<String, AttributeValue> attributes = new HashMap<>();
     attributes.put("delay ms", AttributeValue.longAttributeValue(delay.toMillis()));
     populateError(attributes, error);
 
     String msg = error != null ? "Attempt failed" : "Operation incomplete";
-    span.addAnnotation(msg + ", scheduling next attempt", attributes);
+    attemptCompleted(msg + ", scheduling next attempt", attributes);
   }
 
   /** {@inheritDoc} */
   @Override
   public void attemptFailedRetriesExhausted(Throwable error) {
-    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    Map<String, AttributeValue> attributes = new HashMap<>();
     populateError(attributes, error);
 
-    span.addAnnotation("Attempts exhausted", attributes);
+    attemptCompleted("Attempts exhausted", attributes);
   }
 
   /** {@inheritDoc} */
   @Override
   public void attemptPermanentFailure(Throwable error) {
-    Map<String, AttributeValue> attributes = baseAttemptAttributes();
+    Map<String, AttributeValue> attributes = new HashMap<>();
     populateError(attributes, error);
 
-    span.addAnnotation("Attempt failed, error not retryable", attributes);
+    attemptCompleted("Attempt failed, error not retryable", attributes);
+  }
+
+  private void attemptCompleted(String msg, Map<String, AttributeValue> attributes) {
+    attributes.put("attempt", AttributeValue.longAttributeValue(attemptCount));
+
+    long localAttemptSentMessages = attemptSentMessages.get();
+    if (localAttemptSentMessages > 0) {
+      attributes.put(
+          "attempt request count", AttributeValue.longAttributeValue(localAttemptSentMessages));
+    }
+    if (attemptReceivedMessages > 0) {
+      attributes.put(
+          "attempt response count", AttributeValue.longAttributeValue(attemptReceivedMessages));
+    }
+
+    span.addAnnotation(msg, attributes);
+
+    attemptCount++;
```

Comment:
Why do we do this again? Attempts count should be taken directly from the `TimedAttemptSettings` class (propagated here somehow).

Reply:
It's a little awkward to pass the attemptNumber to all of the attempt finish methods. If you feel strongly about this, then I can compromise on leaving the attemptStarted method to get the attempt # but remove the annotation